receiver/kubeletstatsreceiver/kubelet/client.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubelet
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"go.uber.org/zap"
+)
+
+type AuthType string
+
+const (
+	// AuthTypeTLS indicates that client TLS auth is desired
+	AuthTypeTLS AuthType = "tls"
+	// AuthTypeServiceAccount indicates that the default service account token should be used
+	AuthTypeServiceAccount AuthType = "serviceAccount"
+)
+
+// Config for a kubelet Client. Mostly for talking to the kubelet HTTP endpoint.
+type ClientConfig struct {
+	// AuthType should be "tls" or "serviceAccount". TLS requires that the cert/key
+	// paths below also be set.
+	AuthType AuthType `mapstructure:"auth_type"`
+	// CacertPath should be the path to the root CAs. Used to verify the kubelet
+	// server's certificates.
+	CacertPath string `mapstructure:"ca_cert_path"`
+	// ClientKeyPath should be the path to the TLS client cert. Presented to the
+	// other side of the kubelet connection.
+	ClientCertPath string `mapstructure:"client_cert_path"`
+	// ClientKeyPath should be the path to the TLS client key. Must correspond to
+	// the ClientCertPath, making a key pair to present to the kubelet.
+	ClientKeyPath string `mapstructure:"client_key_path"`
+	// InsecureSkipVerify controls whether the client verifies the server's
+	// certificate chain and host name.
+	InsecureSkipVerify bool `mapstructure:"insecure_skip_verify"`
+}
+
+// A kubelet client, which mostly handles auth when talking to a kubelet server.
+// Marshaling/unmarshaling should be performed by the caller.
+type Client struct {
+	c       *http.Client
+	baseURL string
+	token   string
+	logger  *zap.Logger
+}
+
+func NewClient(baseURL string, cfg *ClientConfig, logger *zap.Logger) (*Client, error) {
+	var client *Client
+	var err error
+	switch cfg.AuthType {
+	case AuthTypeTLS:
+		client, err = tlsClient(baseURL, cfg)
+	case AuthTypeServiceAccount:
+		client, err = serviceAccountClient(baseURL)
+	default:
+		return nil, errors.New("cfg.AuthType not found")
+	}
+	if err != nil {
+		return nil, err
+	}
+	if baseURL == "" {
+		// this requires HostNetwork to be turned on
+		hostname, err := os.Hostname()
```

Comment:
I think it might better for endpoint to be configured explicitly by the end user or via a discovery rule.

Reply:
Sounds good to me. This was taken from SmartAgent BTW.